caledonia/src/bounded.rs

```diff
@@ -0,0 +1,451 @@
+            assert!(params.n_p >= check);
+        }
+
+        Setup {
+            n_p: params.n_p,
+            u,
+            r,
+            d,
+            q,
+            b,
+        }
+    }
+}
+
+/// Round parameters
+#[derive(Debug, Clone)]
+pub struct Round {
+    /// Proof counter
+    v: usize,
+    /// Proof 2nd counter
+    t: usize,
+    // Round candidate tuple
+    s_list: Vec<[u8; 32]>,
+    /// Round candidate hash
+    h: Vec<u8>,
+    /// Round candidate hash mapped to [1, n_p]
+    h_usize: usize,
+    /// Approximate size of set Sp to lower bound
+    n_p: usize,
+}
+
+impl Round {
+    /// Oracle producing a uniformly random value in [1, n_p] used for round candidates
+    /// We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h1(data: Vec<Vec<u8>>, n_p: usize) -> ([u8; 32], usize) {
+        let digest = utils::combine_hashes(data);
+        return (digest, utils::oracle(&digest, n_p));
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub fn new(v: usize, t: usize, n_p: usize) -> Round {
+        let mut data = Vec::new();
+        data.push(v.to_ne_bytes().to_vec());
+        data.push(t.to_ne_bytes().to_vec());
+        let (h, h_usize) = Round::h1(data, n_p);
+        Round {
+            v,
+            t,
+            s_list: Vec::new(),
+            h: h.to_vec(),
+            h_usize,
+            n_p,
+        }
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Round, s: [u8; 32]) -> Round {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let mut data = Vec::new();
+        data.push(r.h.clone());
+        data.push(s.to_vec());
+        let (h, h_usize) = Round::h1(data, r.n_p);
+        Round {
+            v: r.v,
+            t: r.t,
+            s_list,
+            h: h.to_vec(),
+            h_usize,
+            n_p: r.n_p,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
+    /// Proof counter
+    r: usize,
+    /// Proof 2nd counter
+    d: usize,
+    /// Proof tuple
+    items: Vec<[u8; 32]>,
+}
+
+impl Proof {
+    /// Returns an empty proof
+    fn empty() -> Self {
+        Proof {
+            r: 0,
+            d: 0,
+            items: Vec::new(),
+        }
+    }
+
+    /// Oracle producing a uniformly random value in [1, n_p] used for prehashing S_p
+    // TODO: We also return hash(data) to follow the optimization presented in Section 3.3
+    fn h0(setup: &Setup, v: usize, s: [u8; 32]) -> usize {
+        let mut data = Vec::new();
+        data.push(v.to_ne_bytes().to_vec());
+        data.push(s.to_vec());
+        let digest = utils::combine_hashes(data);
+        // return (digest, utils::oracle(&digest, setup.n_p));
```

Comment:
Remove commented code

Reply:
Sure! IIRC, this was a reminder to add the hash of the "s"es to the bins and balls as optimization.